### ISSUE TITLE
Concept: Fleshing out pl-dsdircopy

### ISSUE THREAD

**Issue by mairin** [FNNDSC/ChRIS_ultron_backEnd#231]
# Background:

One of the core concepts in ChRIS is the concept of a "feed." A feed is basically a graph that begins with a single root node. That root node starts off the feed with an initial dataset, and parts of that initial data set are passed down the graph, node-to-node, moving forward. Each plugin has its own container and storage space. And each plugin has one input and one output.

Any given node (plugin) can decide what it passes on to the next node by pushing data to its output, which the next plugin receives through its input. A plugin may pass on all of its filesystem, or it may only pass on a selection of the data it inherited from that initial root node data set. It also of course can pass on new data it generated from its compute. 

## One issue: once a feed has started, new data cannot be added to it outside of what the plugins generate
Once the initial dataset is passed into the root node and the feed has started, there is no other way to pass external data into the feed - new data can be generated by the plugins, but you can't do an operation to add additional data into a node further down the graph once the feed has started.

Where this practically causes an issue - if you have a plugin that requires some kind of external file to operate on a given piece of data - let's say some kind of image filter or mask - and it's the 4th plugin down the line - you could add it to the initial dataset you start the feed with, and have the intervening plugins pass it silently down until it reaches the 4th plugin that will actually use it. I called this the "bucket brigade" problem :) 

The problem: you can't introduce new data mid-stream in a field in the current model. So if you created a feed and forgot when you first set it up (or didn't know until later on you wanted to add X plugin and it has a requirement to have this new data available), you would have to create a new feed, add in that csv field in that initial first node, then recreate the old feed with all the same plugins and settings until you get to that 4th plugin that needs it.

## Another issue: no way to join branched-out nodes

Because the feed has a graph structure, a single node can create branches. For example, you may have a plugin that branches out into three different child nodes on segmented portions of the data. The problem is then if you want to reunify these nodes or take output data from two different nodes in the feed and push it to the input of a new node - you can't. 

## Issue: splitting across too many feeds to achieve this is unwieldy

One way to achieve this linkage today is to split this all across different feeds. You run multiple single-threaded executions each in their own Feeds and then subsequently, once all done, make a new Feed where you combine the results of all these single threaded graphs into the new root. Basicabatching, reproducability, and the overhead of managing itlly a "poor man's" join. The problem for me with that is that you lose the logical encapsulation of an experiment and place a lot more admin burden on an end user and if you have multiple joins and splits of a larger logical workflow, breaking that into multiple single-threads at the Feed level loses so much.

# Requirements

- **batching:** we need to allow users to take the output of multiple nodes and batch it together
- **reproducibility:** any technique or concept we come up with here needs to preserve the reproducibility of any feed
- **minimize management overhead:** our users are researchers who want to focus on creating analyses and consuming their output, they dont want to be in the business of manually orchestrating and managing multiple feeds for one analysis project
- **compatible with ChRIS execution model:** ChRIS doesn't have a model like the one in Tensorflow in which you pre-specify a pipeline (graph) and then it’s executed as a whole - in ChRIS the nodes of the pipelines are too general and users need to experiment one node at a time. It's not worked out upfront. Once they create a pipeline that they know it works then they can publish it as a whole graph (aka tree)

# pl-dsdircopy

There is a plugin called pl-dsdircopy which can accept as input any historical instance (it doesn't even have to be in the same Feed graph) and will combine/copy the contents of all those that it points to into a cumulative outputdir. The "problem" is that this necessitates some UI rethinking since there is a level of meta context now at play. This isn't a "compute" plugin (it doesn't really do any compute) and more of a "topology/graph" plugin. There is no way currently to handle this in the UI (or at least re-create such in the UI).

@rpienaar had the idea for a pl-dsdircopy to record in its outputdir the list of ancestor instances, say ancestors.json and then the UI would have to handle this as a special case in constructing the Feed tree graph

## Challenges

- the backend only has the concept of a single ancestor to a node. I don't think it's possible to change this easily. So when recreating the graphical node graph, a cumulative outputdir that is part of a pl-dsdircopy will only have the immediate ancestor instance in the backend, not the list of N other instances that also contributed to this

# Some example cases to think of

## Example 1 
![Selection_012](https://user-images.githubusercontent.com/799683/91482781-9c0a6a00-e874-11ea-91fb-4394b2641c46.png)
So in the above feed graph, you use pl-dsdircopy as shown to combine the "some analysis" and the "jpg" nodes into one location, which is then consumed by the "report"

UI-wise, the pl-dsdircopy in the BE would only have one ancestor instance (either some analysis or jpg convert) so to recreate/redraw this feed is not possible currently: some additional information needs to exist / be handled. We would need somehow to handle "meta" information in both the UI recreation and also the pipeline running / scheduling / sync etc. If the pl-dsdircopy recorded, when executed, all its ancestors explicitly in some output file, then when the UI (or workflow manager) finds a pl-dsdircopy plugin in a Feed Graph it will have to handle this join operation - in the UI case draw the "connection" between the pl-dsdircopy node and all the ancestors (since that is not available from the BE).

## Example 2

Another case:  when you are needing to combine outputs -- for instance if in a Feed you run 10 FS instances and then want to compute across the space of their outputs then the join becomes critical

Say you have a toy workflow that takes an input data set, say a picture, and say you had some analysis, let's call it a object recognition task, that takes in that input and segments out/counts some label. Say birds cats dogs cars humans - the output of each specific classifier is whatever, say the number of instances of that label. And say the whole point of this was to, for a given picture, count the instances of each label class (and yes, this is a toy and artificial and yada yada yada.) 

it would look something like this:

![Selection_013](https://user-images.githubusercontent.com/799683/91483857-5484dd80-e876-11ea-9a67-b48bbec5bde8.png)

Having that in ONE Feed is superior to having to do this:

![Selection_014](https://user-images.githubusercontent.com/799683/91483906-64042680-e876-11ea-9c5d-b2989c242b8c.png)

And then stop, go back to the Feed create interface and manually select the outputs of all the above in that workflow to make a new Feed:

![Selection_015](https://user-images.githubusercontent.com/799683/91483982-7d0cd780-e876-11ea-9a77-ad488c2ba2f1.png)
Looking at the Feed list, I'd have no real idea that the Feed above is really the result of processing of the original picture

# Example 3 

This is example 2 with multiple pl-dsdircopy nodes:

![Selection_016](https://user-images.githubusercontent.com/799683/91484292-fe646a00-e876-11ea-9fa3-c54bb1fc9844.png)

If you had a report that should just have the counts of cats and dogs you want the graph to show that. You could have several output reports, right? a report for all labels... a report for just mammals, etc.

# The Tasks at Hand

Think through use cases for pl-dsdircopy. The basic case is 5 leaf nodes' output getting cat'ed together / joined via pp-dsdircopy to say generate a report. Each leaf node is processed by the same plugin in the same way.

The more complex cases we have to think through:

1. SCALE - How do we handle joins between 400 leaf nodes? (Say each of M data points contains multiple training sets -- for instance a brain that is segmented into N  different regions. Let's say you wanted to train N different classifiers on each of these regions for the M input subjects. Drawing this all in one Feed makes management and handling very apparent. Handling this at the level of single-threaded only Feeds is a nightmare)

2. MULTIJOIN - How do we handle, of a set of say 20 nodes, multiple different joins via pl-dsdircopy (eg, nodes 2, 4, and 6 are group 1. nodes 1, 7, 3, and 5 are group 2. Etc.)

3. HETEREOGENEOUS - What if we're joining output that is from different branches of the same tree / generated from diff plugins with different settings?

Some other concepts to think about:

1. How do you address / select nodes in a scaleable way across a set of say 400? Do you use some regexp like system to build filters?

2. How do you address new data coming in and being needed to add to the join? Does this require a new feed, or can it be added?

3. Do we support cross feed joins?

## UI Considerations

1.  the UI will need to have some UI-specific context logic. If the UI sees pl-dsdircopy it will have to know that the ancestor that the BE told it is not the whole picture. The complete picture knowledge has to be found elsewhere. This could be in the values of the CLI passed to that node. It could be a special file in the pl-dsdircopy that is written that the FE, when re-drawing this will collect and parse.
1.  the UI will need some additional FE "drawing" logic. When you choose "Add a new node" and if you choose pl-dsdircopy then somehow the UI should know "whoa. Ok. there's a join coming" and has to go into some mode that allows the user to, I dunno, click on nodes in the tree and as they are clicked it "draws" the join? 
1. How closely to have the feed graph represent/mirror logic/computation? You ultimately reduce all this always to single threaded Feed flow with complex patterns of CLI etc. For example, if multiple reports are generated in example 3, visualize the data set as one box with multiple report nodes coming off of it.

**Comment by junaruga** [comment]
If we can summarize, this ticket is about a new feature to support multiple ancestors, right?

**Comment by mairin** [comment]
@junaruga Yep I think that'd be accurate!

**Comment by mairin** [comment]
@willianrampazzo Any updates on this?

**Comment by willianrampazzo** [comment]
> @willianrampazzo Any updates on this?

No updates yet, sorry.

**Comment by mairin** [comment]
No prob, just realized we hadn't checked in for a while :)

**Comment by jbernal0019** [comment]
A new category of plugins has been added to ChRIS. These are known as "topological" plugins (`ts` plugins). They can be used to join many branches within a feed and to split a single node into many branches. This is usually achieved with `pl-topologicalcopy` plugin. 

For the use  case of adding an additional file mid-way a feed users can use `pl-tsdircopy` or even `pl-dsdircopy`. However the first (`ts`) is more efficient as it doesn't send the entire output directory of the previous plugin to the remote compute to just add a new file that isn't even in that output directory. `pl-tsdircopy` also includes all the functionality of `pl-topologicalcopy`. 

The only advantage of `pl-dsdircopy` is that it could potentially be used as part of a **backend pipeline** (an entire tree of plugins that is sent as a batch of jobs to a remote compute environment). Backend pipelines can only be comprised of `ds` plugins.  However they remote compute services (`pfcon` and `pman`) can not yet deal with backend pipelines. This is in the future roadmap.

Also @PintoGideon has already implemented the UI processes to deal with the above joins and splits.

So @mairin can we close this issue?

**Comment by rudolphpienaar** [comment]
I’m fine with closing this

[Rudolph Pienaar - Chat @ Spike](https://spikenow.com/r/a/?ref=spike-organic-signature&_ts=17d3be)	[17d3be]

On September 13, 2021 at 16:24 GMT, Jorge ***@***.***> wrote:

A new category of plugins has been added to ChRIS. These are known as "topological" plugins (ts plugins). They can be used to join many branches within a feed and to split a single node into many branches. This is usually achieved with pl-topologicalcopy plugin.

For the use case of adding an additional file mid-way a feed users can use pl-tsdircopy or even pl-dsdircopy. However the first (ts) is more efficient as it doesn't send the entire output directory of the previous plugin to the remote compute to just add a new file that isn't even in that output directory. pl-tsdircopy also includes all the functionality of pl-topologicalcopy.

The only advantage of pl-dsdircopy is that it could potentially be used as part of a backend pipeline (an entire tree of plugins that is sent as a batch of jobs to a remote compute environment). Backend pipelines can only be comprised of ds plugins. However they remote compute services (pfcon and pman) can not yet deal with backend pipelines. This is in the future roadmap.

Also ***@***.***(https://github.com/PintoGideon) has already implemented the UI processes to deal with the above joins and splits.

So ***@***.***(https://github.com/mairin) can we close this issue?

**Comment by jbernal0019** [comment]
OK, going ahead and closing this issue!
